### ISSUE TITLE
Announcing WinUI 3 Preview 4! 🎉

### ISSUE THREAD

**Issue by anawishnoff** [microsoft/microsoft-ui-xaml#4156]
# WinUI 3 Preview 4 

We've just released WinUI 3 Preview 4! This release consists of stabilization improvements, critical bug fixes, and some new functionality. 

We expect to ship the first stable, supported version of WinUI 3 in March. 

### Here are some capabilities that are new since Preview 3:

- Parity with WinUI 2.5 (includes InfoBar control, new features in ProgressRing and NavigationView, and bug fixes)
- Custom titlebar capabilities: new Window.ExtendsContentIntoTitleBar and Window.SetTitleBar APIs that allow developers to create custom title bars in Desktop apps. 
- VirtualSurfaceImageSource  support

### And here are just a few of the critical bug fixes included in this release:

- WinUI 3 crash when running on Windows Insider Builds
- WinUI 3 Desktop: Unable to Enter Text into TextBox inside ContentDialog 
- WinUI 3 Desktop: DataGrid not displaying
- This release has taken on a new version of CS/WinRT and the Windows SDK, which fixed the following bugs:
  - Crash when binding to a URI property using {Binding}
  - C#/WinRT Marshal functions not interoperating correctly with .NET 5
- WinUI 3 Desktop: Attempting to bring down DropDownButton or SplitButton's flyout causes app crash

For a full list of the bugs fixed in this release, head over to our [release notes](https://aka.ms/winui3/preview4). If you filed a bug on this repo that was fixed in this release, your issue will be closed shortly. 

Please keep in mind that Preview 4 is still a pre-release product, and has a number of [known issues and limitations](https://aka.ms/winui3/preview4#Limitations-and-known-issues) that our team continues to work on. This preview is not intended for production apps.  

## Getting started with Preview 4
First, you'll need to set up your dev environment with the appropriate technologies. See our documentation for installation instructions and more information on project types: 

**[Install WinUI 3 Preview 4](https://aka.ms/winui3/preview4#Install-WinUI-3-Preview-4)**

API Reference documentation for Preview 4 is coming *soon*, and this issue will be updated when it's available. 

Take a look at the walk-through documentation on [building a UWP app with WinUI 3 Preview 4](https://aka.ms/winui3/uwpguide) and [building a Desktop app with WinUI 3 Preview 4](https://aka.ms/winui3/desktopguide). 

We love feedback! We encourage you to file any bug, big or small, on our repo using [this template](https://github.com/microsoft/microsoft-ui-xaml/issues/new?assignees=&labels=&template=bug_report.md&title=). Knowing which issues and features are important and/or critical to our customers highly influences which ones we tackle.  

For upgrading Preview 3 apps to Preview 4, carefully follow the instructions laid out [here](https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/preview_upgrade_instructions.md). 

## Using WinUI 3 Preview 4 with our ecosystem partners

Preview 4 is in the process of being integrated into several other important ecosystem technologies. To test out and follow that progress, check out some of our partner technologies below:

- [Windows Community Toolkit (Microsoft)](https://aka.ms/wct-winui3): The latest version of the WCT supports Preview 4, and provides tons of new controls and capabilities for use in your WinUI app. You can check out their open source repo [here](https://aka.ms/windowstoolkit).

- [Uno Platform](https://github.com/unoplatform): Developers using WinUI have an option to take WinUI-built applications to WebAssembly, Linux, macOS, Android and iOS via open-source project Uno Platform, and easily style them with Material and Fluent UI design systems for pixel-perfect experience everywhere. To set up your environment and get started with Uno Platform follow instructions laid out [here](https://platform.uno/docs/articles/getting-started-tutorial-1.html). 

- [Telerik](https://www.telerik.com/winui): Telerik UI for WinUI is a UI component suite targeting WinUI 3 UWP apps currently released as Preview. The Telerik UI Preview includes several feature-rich controls including Ribbon, DataGrid, Charts, Gauges, Barcode, and more.

- [Syncfusion](https://www.syncfusion.com/winui-controls): The Syncfusion WinUI controls preview includes a large variety of high performance and modern controls for WinUI 3 UWP apps. The initial preview includes a suite of controls for data visualization, notifications, navigation, and content editing.

- [DevExpress](https://community.devexpress.com/blogs/wpf/archive/2020/06/11/devexpress-controls-for-winui-3-preview-are-now-available.aspx): The entire suite of DevExpress UWP controls is available in preview for WinUI 3 UWP apps. Some example controls include Scheduler, Charts, Navigation, DataGrid, and more  

- [GrapeCity's service components](https://www.grapecity.com/componentone/net-service-component) allow users to integrate data from popular online data sources such as Dynamics, Salesforce, Quickbooks and Google Analytics, and optimize performance with data virtualization. GrapeCity also plans to bring their popular desktop UI controls to WinUI3 later this year.

- [Infragistics](https://www.infragistics.com/products/winui): Ultimate UI for WinUI brings business critical, high performing, and feature rich line of business controls to your apps that target any platform that runs Windows, including the new ARM64-based Surface Pro X. The Ultimate UI or WinUI Preview includes DataGrid, Charts (50+), Gauges, and more.

## What's next
To keep up with progress being made on WinUI 3, please see our [feature roadmap](https://aka.ms/winui3/feature-roadmap) which gets updated regularly. We also give monthly updates on our WinUI Community Calls, where you can directly ask the team any questions you have.

**Comment by brianlagunas** [comment]
[Infragistics](https://www.infragistics.com/products/winui): Ultimate UI for WinUI brings business critical, high performing, and feature rich line of business controls to your apps that target any platform that runs Windows, including the new ARM64-based Windows 10X. The Ultimate UI or WinUI Preview includes DataGrid, Charts (50+), Gauges, and more.

Be sure to [read our blog](https://www.infragistics.com/community/blogs/b/infragistics/posts/introducing-the-ultimate-ui-for-winui-preview)

Install via [NuGet](https://www.nuget.org/packages?q=Infragistics.WinUI)

**Comment by anawishnoff** [comment]
Thanks @brianlagunas - I'll add your description to the list! 😊

**Comment by vitorgrs** [comment]
What's the plan for MediaPlayer control?

**Comment by jeromelaban** [comment]
If you're getting this error when building an app:

```
error CS1705: Assembly 'Microsoft.WinUI'
with identity 'Microsoft.WinUI, Version=3.0.0.0, Culture=neutral, PublicKeyToken=de31ebe4ad15742b' uses
'WinRT.Runtime, Version=1.1.0.0, Culture=neutral, PublicKeyToken=99ea127f02d97709' which has a higher
version than referenced assembly 'WinRT.Runtime' with identity 'WinRT.Runtime, Version=1.0.1.33301,
Culture=neutral, PublicKeyToken=99ea127f02d97709'
```

Make sure to update to the latest .NET 5 SDK (5.0.100 fails and 5.0.103 fixes it).

**Comment by robloo** [comment]
@vitorgrs 

> What's the plan for MediaPlayer control?

"Planned for a future update" https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/roadmap.md#winui-30-feature-roadmap So the plan is there is a plan. Which is a slightly better situation than `MapControl`. The lack of clarity in my response is intentional.

**Comment by LeftTwixWand** [comment]
Is WinUI 3 planing to be open source?

**Comment by chingucoding** [comment]
@LeftTwixWand Yes, WinUI 3 will be open source, however this will after GA release.
@anawishnoff @ryandemopoulos Do you think it's a good thing to add that to the roadmap?

**Comment by LeftTwixWand** [comment]
@chingucoding Thank you! I'm asking, because I want to change the style of a component, using: Edit Template -> Edit a copy (Or Edit current), but now I can't to do it.

**Comment by chingucoding** [comment]
@LeftTwixWand You might be able to grab the template from generic.xaml or from this repository if it's a WinUI 2 control such as NumberBox or Expander. But the team might now more what the best workaround for this is.

**Comment by EricOuellet2** [comment]
Where can we know the status of features like SVG (read, interraction, convert to WinUI graphic objects for interraction), DataGrid, Docking window, Ribbon menu, ...?
Where can we post our feature request in order for you to see the most requested ones and for us to vote for any one or create new ones?

**Comment by EricOuellet2** [comment]
Where can we know the status of features like SVG (read, interraction, convert to WinUI graphic objects for interraction), DataGrid, Docking window, Ribbon menu, ...?
Where can we post our feature request in order for you to see the most requested ones and for us to vote for any one or create new ones?

**Comment by anawishnoff** [comment]
@EricOuellet2 You can open a feature request issue on this repo, or open a discussion or question issue and ask for the status of these features, as well as request that they be added to the feature roadmap. A quicker path to get answers may be to ask about these features in the next Community Call - you can find the GitHub issue for yesterday's community call here with more info on that: #4150

**Comment by EricOuellet2** [comment]
@anawishnoff Thank you for your quick feedback. For reporting and following bug, Github is great. But it is not really great for feature request - I mean nobody could manage each request individually if they are already been asked and/or answered.

**Comment by nlogozzo** [comment]
Is there support for using community controls such as a data grid with C++ Win32 apps and WinUi3. Many apps are useless without data grids...

**Comment by anawishnoff** [comment]
@nlogozzo Unfortunately there's no built-in DataGrid option in WinUI 3 for C++ apps. For Win32 C#/.NET apps, you can use the [Windows Community Toolkit 8.0 Preview 4](https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3295), which is an add-on package that includes a DataGrid control. 

I would also recommend taking a look at the ecosystem partners mentioned in the announcement post above - there are a few control vendors listed that may have what you're looking for.

**Comment by EricOuellet2** [comment]
@anawishnoff Thanks a lot for you feedback on nlogozzo. You really makes WinUI easier to use. You do a great job. But I feel to product is far from being ready and I also think Microsoft should use their own product if they want peoples to trusts them. Do you think Elon Musk drive a GM or a Chrysler? Will he be credible to do so? The day Microsoft will begin to use C# to develop their own product will clearly say the world they really should use it!!! And I easily guaranty a great success.

**Comment by nlogozzo** [comment]
> @nlogozzo Unfortunately there's no built-in DataGrid option in WinUI 3 for C++ apps. For Win32 C#/.NET apps, you can use the [Windows Community Toolkit 8.0 Preview 4](https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3295), which is an add-on package that includes a DataGrid control. 
> 
> 
> 
> I would also recommend taking a look at the ecosystem partners mentioned in the announcement post above - there are a few control vendors listed that may have what you're looking for.

It looks like all those vendors support only C#. I have C++ code that I would like to add a Win32 UI too easily and WinUi 3 provides that but I need a data grid...it'd love to see a WinUI 3 control for one

**Comment by shelllet** [comment]
I can't imagine that the offical version will be released next month, because desktop c++/winrt have many problems.

**Comment by gabbybilka** [comment]
Check out #6172 to see features and bug fixes from this release in Windows App SDK 1.0.
